redis/src/cluster_async/mod.rs

```diff
@@ -1220,30 +1232,32 @@ where
             return;
         }
 
-        // validate active subscriptions location
         let mut addrs_to_refresh: HashSet<ArcStr> = HashSet::new();
-        // lock the subscriptions and connections for writing - might be possible to optimize
         let mut subs_by_address_guard = inner.subscriptions_by_address.write().await;
         let mut unassigned_subs_guard = inner.unassigned_subscriptions.write().await;
+        let conns_read_guard = inner.conn_lock.read().await;
 
-        let mut conns_write_guard = inner.conn_lock.write().await;
-
+        // validate active subscriptions location
         subs_by_address_guard.retain(|addr, addr_subs| {
             addr_subs.retain(|kind, channels_patterns| {
                 channels_patterns.retain(|channel_pattern| {
                     let new_slot = get_slot(channel_pattern);
                     let mut valid = false;
-                    if let Some((new_address, _)) = conns_write_guard
+                    let mut address_exists = false;
+                    if let Some((new_address, _)) = conns_read_guard
                         .connection_for_route(&Route::new(new_slot, SlotAddr::Master))
                     {
+                        address_exists = true;
```

Comment:
address_exists shouldn't refer to if the old address exists? here you're checking if the new address exists

Reply:
fixed